net/net processing: Move addr data into net_processing

jnewbery opened the pull request:
This continues the work of moving application layer data into net_processing, by moving all addr data into the new Peer object added in #19607.

For motivation, see #19398.

jnewbery committed:
scripted-diff: rename address relay fields

-BEGIN VERIFY SCRIPT-
ren() { sed -i "s:\<$1\>:$2:g" $(git grep -l "\<$1\>" ./src ./test); }

ren fGetAddr     m_getaddr_sent
ren fSentAddr    m_getaddr_recvd
ren vAddrToSend  m_addrs_to_send
-END VERIFY SCRIPT-

jnewbery committed:
[net processing] Move addr relay data and logic into net processing

jnewbery committed:
[net processing] Take NodeId instead of CNode* as originator for RelayAddress()

This makes the following commit easier.

jnewbery committed:
[refactor] Remove unused ForEachNodeThen() template

jnewbery committed:
[net processing] Make RelayAddress() a member function of PeerManagerImpl

fanquake commented on src/net_processing.cpp:
```diff
@@ -216,7 +237,8 @@ struct Peer {
 
-    explicit Peer(NodeId id) : m_id(id) {}
+    explicit Peer(NodeId id, bool addr_relay)
+        : m_id(id), m_addr_known{addr_relay ? nullptr : MakeUnique<CRollingBloomFilter>(5000, 0.001)} {}
```
Please [use `std::make_unique` in new code](https://github.com/bitcoin/bitcoin/blob/master/src/util/memory.h#L13).

jnewbery commented:
#21236 is merged. Moving this out of draft.

hebasto commented:
Concept ACK.

hebasto commented on src/net_processing.cpp:
```diff
@@ -218,7 +241,10 @@ struct Peer {
     std::deque<CInv> m_getdata_requests GUARDED_BY(m_getdata_requests_mutex);
 
-    explicit Peer(NodeId id) : m_id(id) {}
+    explicit Peer(NodeId id, bool addr_relay)
+        : m_id(id)
+        , m_addr_known{addr_relay ? nullptr : std::make_unique<CRollingBloomFilter>(5000, 0.001)}
+    {}
```
Out of curiosity, is there a way to teach the [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy) to support such formatting, i.e., both a leading comma and `{}` on the the same line?

jnewbery replied:
Good question! I'm not an expert on clang-format.

jnewbery replied:
Perhaps `BCIS_BeforeComma`? (https://clang.llvm.org/docs/ClangFormatStyleOptions.html)

hebasto commented on src/net_processing.cpp:
```diff
@@ -1483,59 +1551,49 @@ void PeerManagerImpl::RelayTransaction(const uint256& txid, const uint256& wtxid
     unsigned int nRelayNodes = (fReachable || (hasher.Finalize() & 1)) ? 2 : 1;
 
-    std::array<std::pair<uint64_t, CNode*>,2> best{{{0, nullptr}, {0, nullptr}}};
+    std::array<std::pair<uint64_t, Peer*>,2> best{{{0, nullptr}, {0, nullptr}}};
```
style nit:
```suggestion
    std::array<std::pair<uint64_t, Peer*>, 2> best{{{0, nullptr}, {0, nullptr}}};
```

jnewbery replied:
Fixed!

hebasto commented on src/net_processing.cpp:
```diff
@@ -2674,24 +2732,24 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             }
             bool fReachable = IsReachable(addr);
-            if (addr.nTime > nSince && !pfrom.fGetAddr && vAddr.size() <= 10 && addr.IsRoutable())
+            if (addr.nTime > nSince && !peer->m_getaddr_sent && vAddr.size() <= 10 && addr.IsRoutable())
             {
```
style nit:
```suggestion
            if (addr.nTime > nSince && !peer->m_getaddr_sent && vAddr.size() <= 10 && addr.IsRoutable()) {
```

jnewbery replied:
Done

hebasto reviewed (commented):
Approach ACK 468db1469382ee11380f0667454b86ec1903eca8.

In _"[net processing] Move addr relay data and logic into net processing"_ commit (6fdf4b62e3161ce6fdd27b778de53933dccde65e) four functions are moved from `CNode` members to the `net_processing.cpp` as free `static` functions. I agree that there's no need to make them members of `Peer`, but why not place them in a namespace? It could be unnamed, or we could choose a pretty good name for it :)

Also `CNode::AddAddressKnown` and `CNode::PushAddress` are subjects for fuzzing in the master branch. Could new free functions also be fuzzed?

jnewbery commented:
Thanks for the review, @hebasto! I've addressed both of your style comments.

> four functions are moved from CNode members to the net_processing.cpp as free static functions. I agree that there's no need to make them members of Peer, but why not place them in a namespace? It could be unnamed, or we could choose a pretty good name for it :)

There's almost no difference between declaring a free function `static` and placing it in the unnamed namespace. Both will prevent the name from being accessible outside the translation unit. See https://stackoverflow.com/a/154482/933705 for example.

> Also CNode::AddAddressKnown and CNode::PushAddress are subjects for fuzzing in the master branch. Could new free functions also be fuzzed?

Both of these functions are called from `ProcessMessage()`, so are indirectly tested by the `fuzz/process_messages.py` fuzzer.

hebasto reviewed (approved):
ACK c538369123b6c69358649255815cf331b2c39f52, I have reviewed the code and it looks OK, I agree it can be merged.

> There's almost no difference between declaring a free function `static` and placing it in the unnamed namespace. Both will prevent the name from being accessible outside the translation unit. See https://stackoverflow.com/a/154482/933705 for example.

https://github.com/bitcoin/bitcoin/pull/19176#discussion_r435937252

jnewbery commented:
> #19176 (comment)

I don't see a hugely compelling reason to prefer unnamed namespaces over static declaration in any of those links, but equally wouldn't be opposed to updating the style guide to prefer unnnamed namespaces just for consistency.

One advantage of using `static` that isn't mentioned is that it's instantly obvious to the reader that the symbol has internal linkage. Unnamed namespaces often span many hundreds of lines, so it's not obvious when something is in the namespace. For example, I've just realized that in this PR, all the new functions are in fact inside an unnamed namespace that spans lines 542-942.

hebasto commented:
> For example, I've just realized that in this PR, all the new functions are in fact inside an unnamed namespace that spans lines 542-942.

There are three adjacent unnamed namespaces...

> Unnamed namespaces often span many hundreds of lines...

I think it is a code organizing problem, no?

jnewbery commented:
> There are three adjacent unnamed namespaces...

Yes. This is part of the unfinished work in #20758. Once that's finished, these unnamed namespaces should be fixed up.

> > Unnamed namespaces often span many hundreds of lines...

> I think it is a code organizing problem, no?

I'm not sure. I think if we use unnamed namespaces for all the internal functions that we don't want exposed outside the translation unit then they'll usually be hundreds of lines.

hebasto commented:
> I'm not sure. I think if we use unnamed namespaces for all the internal functions that we don't want exposed outside the translation unit then they'll usually be hundreds of lines.

Correct. I mean that our code base have some really huge translation units :)

mzumsande reviewed (commented):
Concept ACK to moving addr-related logic to net_processing. See my other comments for some possible issues.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3566,20 +3622,20 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         // Making nodes which are behind NAT and can only make outgoing connections ignore
         // the getaddr message mitigates the attack.
-        if (!pfrom.IsInboundConn()) {
+        if (!RelayAddrsWithPeer(*peer)) {
```
oh wow. That's terrible. Thank you for catching this.

This was a bad rebase on top of 49c10a9ca40, which combined the two error conditions:

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 690b59476b..859b67755d 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -3521,11 +3521,7 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
         // Making nodes which are behind NAT and can only make outgoing connections ignore
         // the getaddr message mitigates the attack.
         if (!pfrom.IsInboundConn()) {
-            LogPrint(BCLog::NET, "Ignoring \"getaddr\" from outbound connection. peer=%d\n", pfrom.GetId());
-            return;
-        }
-        if (!pfrom.RelayAddrsWithConn()) {
-            LogPrint(BCLog::NET, "Ignoring \"getaddr\" from block-relay-only connection. peer=%d\n", pfrom.GetId());
+            LogPrint(BCLog::NET, "Ignoring \"getaddr\" from %s connection. peer=%d\n", pfrom.ConnectionTypeAsString(), pfrom.GetId());
             return;
         }
```

mzumsande replied:
Oh, I see. There should be a functional test for this that would have failed - something which I [once](https://github.com/bitcoin/bitcoin/pull/19315#issuecomment-701571168) wanted to do anyway but then forgot about. I'll  write one soon!

jnewbery replied:
I'll very happily review once you open the PR.

amitiuttarwar replied:
heads up that I've recently written some similar tests here: https://github.com/bitcoin/bitcoin/pull/21528/files?file-filters%5B%5D=.py#diff-5a9f30817894348260893f195cea07da88271b5e01b1531f0cd839ee85898592R150. I don't think it would cover this exact use case, but just mentioning incase its useful / incase the test improvements could play nice together.

mzumsande replied:
Cool - I didn't know, I had only looked at the non-test commits in your PR! This use case is the reverse direction (inbound/outbound send the getaddr to our node and get back addrs/nothing) from your tests, so I agree that these belong together.

jnewbery replied:
More tests added in #21707

jnewbery commented on src/net_processing.cpp:
```diff
@@ -218,7 +241,10 @@ struct Peer {
     std::deque<CInv> m_getdata_requests GUARDED_BY(m_getdata_requests_mutex);
 
-    explicit Peer(NodeId id) : m_id(id) {}
+    explicit Peer(NodeId id, bool addr_relay)
```
You're right. The logic here is correct, but the naming is confusing. I've reversed the logic, and added comments to the place where this ctor is called to hopefully make this more intuitive.

jnewbery commented on src/net.h:
```diff
@@ -657,37 +632,6 @@ class CNode
-        // SendMessages will filter it again for knowns that were added
-        // after addresses were pushed.
-        assert(m_addr_known);
-        if (_addr.IsValid() && !m_addr_known->contains(_addr.GetKey()) && IsAddrCompatible(_addr)) {
```
This is another rebase error. Thank you for catching!

jnewbery commented:
Thank you for the careful review @mzumsande. You caught two rebase errors :flushed: 

I've fixed those and addressed your other comment. I've also made some minor edits to comments to clarify concepts, and also re-reviewed everything to make sure no other errors have crept in.

ajtowns reviewed (commented):
The "Move addr relay data and logic into net processing" commit is doing a lot of things, and I think that makes it a bit hard to review (and leads to rebase errors?). Might be worth putting a bit of effort into splitting it up further if some of the changes can be isolated?

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -626,6 +650,39 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+    peer.m_addr_known->insert(addr.GetKey());
+}
+
+static void PushAddress(Peer& peer, const CAddress& addr, FastRandomContext& insecure_rand)
```
I'm curious- why did you opt to make this a standalone function that takes in a peer vs a member on peer itself (like it previously was on CNode)?

jnewbery replied:
Because `Peer` isn't a class. It's a data structure containing information about a Peer, which PeerManager acts on.

I know the distinction is blurry in C++, but I think it's generally better to treat things as purely a data structure (which carries data that is acted on by external logic) or purely a class (where the data is private and internal logic exists behind well-defined public interface functions). Mixing the two, where some logic is contained within the object but external logic can also reach into the object and manipulate the data, is confusing.

amitiuttarwar reviewed (commented):
started reviewing, looks good so far, will be awesome to have addr handling in net processing :)

jnewbery commented on src/net_processing.cpp:
```diff
@@ -210,6 +212,25 @@ struct Peer {
+    /** Time point to possibly re-announce our local address to this peer. */
+    std::chrono::microseconds m_next_local_addr_send GUARDED_BY(m_addr_send_times_mutex){0};
+    /** Whether the peer has signaled support for receiving ADDRv2 (BIP155)
+     *  messages, implying a preference to receive ADDRv2 instead of ADDR ones. */
```
Fixed!

jnewbery commented on src/net_processing.cpp:
```diff
@@ -150,6 +150,8 @@ static constexpr uint32_t MAX_GETCFILTERS_SIZE = 1000;
 static constexpr uint32_t MAX_GETCFHEADERS_SIZE = 2000;
 /** the maximum percentage of addresses from our addrman to return in response to a getaddr message. */
 static constexpr size_t MAX_PCT_ADDR_TO_SEND = 23;
+/** The maximum number of new addresses to relay to a peer in an ADDR message. */
```
Much better. Thanks.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4147,72 +4202,72 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer, std::chrono::mic
+    if (!RelayAddrsWithPeer(peer)) return;
 
-    assert(node.m_addr_known);
+    assert(peer.m_addr_known);
```
Good point. Removed!

jnewbery commented on src/net_processing.cpp:
```diff
@@ -617,6 +652,39 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
     return &it->second;
 }
 
+static bool RelayAddrsWithPeer(const Peer& peer) { return peer.m_addr_known != nullptr; }
```
Done.

jnewbery commented:
Thanks for the reviews @ajtowns and @amitiuttarwar. I've addressed all of your comments.

> The "Move addr relay data and logic into net processing" commit is doing a lot of things, and I think that makes it a bit hard to review (and leads to rebase errors?). Might be worth putting a bit of effort into splitting it up further if some of the changes can be isolated?

I did try that at one point, but it seemed more confusing/difficult than just moving across the addr data/logic in one commit.

practicalswift commented:
Concept ACK

laanwj commented:
Code review ACK 0829516d1f3868c1c2ba507feee718325d81e329

jonatack commented:
Concept ACK

mzumsande commented:
ACK 0829516d1f3868c1c2ba507feee718325d81e329, reviewed the code and ran tests.

sipa reviewed (commented):
utACK 0829516d1f3868c1c2ba507feee718325d81e329

jnewbery commented on src/net_processing.cpp:
```diff
@@ -954,7 +1014,9 @@ void PeerManagerImpl::InitializeNode(CNode *pnode)
-        PeerRef peer = std::make_shared<Peer>(nodeid);
+        // Addr relay is disabled for outbound block-relay-only peers to
+        // prevent adversaries from inferring these links from addr traffic.
+        PeerRef peer = std::make_shared<Peer>(nodeid, /* addr_relay = */ !pnode->IsBlockOnlyConn());
```
This logic of checking `IsBlockOnlyConn()` is used only when calling the `Peer` ctor. After that, we use the `RelayAddrsWithPeer()`, which replaces the old `RelayAddrsWithConn()` function. That's a clearer separation between net and net_processing since we only check the connection type when initializing the Peer object.

hebasto reviewed (approved):
re-ACK 0829516d1f3868c1c2ba507feee718325d81e329
